### PR TITLE
Make Database a mutable API that vends immutable Datasets

### PR DESCRIPTION
**Pull request by cmasone-attic**
Noms SDK users frequently shoot themselves in the foot because they're
holding onto an "old" Database object. That is, they have a Database
tucked away in some internal state, they call Commit() on it, and
don't replace the object in their internal state with the new Database
returned from Commit.

This PR changes the Database and Dataset Go API to be in line with the
proposal in Issue #2589. JS follows in a separate patch.

### CODE REVIEW

**Comment by cmasone-attic** [comment]
PTAL, @arv. It's probably easiest just to talk through this patch. The majority of changed lines are just updating callsites. The meat is in the `datas` package (clearly) and there's also a bit of a change to `spec`.

File: go/datas/dataset.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+// Dataset is a named Commit within a Database.
+type Dataset struct {
+	store   Database
+	id      string
+	headRef types.Ref
+}
+
+// Database returns the Database object in which this Dataset is stored.
+func (ds Dataset) Database() Database {
```
Comment (arv) [inline go/datas/dataset.go]: Add a (deprecation) warning here?

File: go/datas/database.go [hunk 2]
```diff
@@ -13,37 +13,87 @@ import (
...
 	types.ValueReadWriter
 	io.Closer
 
-	// MaybeHead returns the current Head Commit of this Database, which contains the current root of the Database's value tree, if available. If not, it returns a new Commit and 'false'.
-	MaybeHead(datasetID string) (types.Struct, bool)
-
-	// MaybeHeadRef returns the types.Ref of the Head Commit of this Database, and true, if available. If not, it returns an invalid types.Ref and false.
-	MaybeHeadRef(datasetID string) (types.Ref, bool)
+	// Datasets returns the root of the database which is a
+	// MapOfStringToRefOfCommit where string is a datasetID.
```
Comment (cmasone-attic) [inline go/datas/database.go]: done

File: go/datas/database.go [hunk 3]
```diff
@@ -13,37 +13,87 @@ import (
...
-	Head(datasetID string) types.Struct
+	// GetDataset returns a Dataset struct containing the current mapping of
+	// datasetID in the above Datasets Map.
+	GetDataset(datasetID string) Dataset
 
-	// HeadRef returns the ref of the current head Commit. See Head(datasetID).
-	HeadRef(datasetID string) types.Ref
+	// Commit updates the Commit that ds.ID() in this database points at. All
+	// Values that have been written to this Database are guaranteed to be
+	// persistent after Commit().
```
Comment (cmasone-attic) [inline go/datas/database.go]: done

File: go/datas/database.go [hunk 4]
```diff
@@ -13,37 +13,87 @@ import (
...
+	// The new Commit struct is constructed using `v`, `opts.Parents`, and
+	// `opts.Meta`. If `opts.Parents` is the zero value (`types.Set{}`) then
+	// the  current head is used. If `opts.Meta is the zero value
+	// (`types.Struct{}`) then a fully initialized empty Struct is passed to
+	// NewCommit.
+	// If the update cannot be performed, e.g.,  because of a conflict, Commit
```
Comment (cmasone-attic) [inline go/datas/database.go]: done

File: go/datas/database.go [hunk 5]
```diff
@@ -13,37 +13,87 @@ import (
...
+	// returns an 'ErrMergeNeeded' error and the  current snapshot of the
+	// Dataset so that the client can merge the  changes and try again.
+	// Regardless, Datasets() is updated to match  backing storage upon return.
+	Commit(ds Dataset, v types.Value, opts CommitOptions) (Dataset, error)
```
Comment (cmasone-attic) [inline go/datas/database.go]: took another pass

File: go/datas/database.go [hunk 6]
```diff
@@ -13,37 +13,87 @@ import (
...
+	Delete(ds Dataset) (Dataset, error)
 
-	// Delete removes the Dataset named datasetID from the map at the root of the Database. The Dataset data is not necessarily cleaned up at this time, but may be garbage collected in the future. If the update cannot be performed, e.g., because of a conflict, error will non-nil. The newest snapshot of the database is always returned.
-	Delete(datasetID string) (Database, error)
+	// SetHead ignores any lineage constraints (e.g. the current Head being in
+	// commit’s Parent set) and force-sets a mapping from datasetID: commit in
+	// this database.
+	// All Values that have been written to this Database are guaranteed to be
+	// persistent after SetHead(). If the update cannot be performed, e.g.,
+	// because of a conflict, error will be non-nil.
```
Comment (cmasone-attic) [inline go/datas/database.go]: done

File: go/datas/database.go [hunk 7]
```diff
@@ -13,37 +13,87 @@ import (
...
+	SetHead(ds Dataset, newHeadRef types.Ref) (Dataset, error)
 
-	// SetHead sets the Commit that datasetID in this database points at. All Values that have been written to this Database are guaranteed to be persistent after SetHead(). If the update cannot be performed, e.g., because of a conflict, error will be non-nil. The newest snapshot of the database is always returned.
-	SetHead(datasetID string, commit types.Struct) (Database, error)
+	// FastForward takes a types.Ref to a Commit object and makes it the new
+	// Head of ds iff it is a descendant of the current Head. Intended to be
+	// used e.g. after a call to Pull(). If the update cannot be performed,
+	// e.g., because another process moved the current Head out from under
+	// you, err will be non- nil.
+	// The newest snapshot of the Dataset is always returned, so the caller an
```
Comment (cmasone-attic) [inline go/datas/database.go]: done

File: go/datas/database_common.go [hunk 8]
```diff
@@ -27,84 +27,89 @@ var (
...
-func (ds *databaseCommon) HeadRef(datasetID string) types.Ref {
-	r, ok := ds.MaybeHeadRef(datasetID)
-	d.PanicIfFalse(ok, "Database \"%s\" has no Head.", datasetID)
-	return r
-}
-
-func (ds *databaseCommon) Datasets() types.Map {
-	if ds.datasets == nil {
-		if ds.rootRef.IsEmpty() {
+func (dbc *databaseCommon) Datasets() types.Map {
```
Comment (cmasone-attic) [inline go/datas/database_common.go]: I'm surprised there isn't.

In this case, though, this type isn't exported, so the comment wouldn't show up in godoc

File: go/datas/dataset.go (hunk 1 above)
Comment (cmasone-attic) [inline go/datas/dataset.go]: done

**Comment by arv** [comment]
I just realized:

> I find `dataset.Commit` cleaner and I would not mind removing `database.Commit`.

Was that considered?

**Comment by arv** [comment]
Talked it through. Commit is updating the database so it belongs on the database.

**Pull request merged**